src/asm2wasm.h

```diff
@@ -1980,9 +1979,17 @@ Function* Asm2WasmBuilder::processFunction(Ref ast) {
           specific->type = fullType->result;
         }
         if (import) {
+          // apply the detected type from the parent
+          // note that this may not be complete, e.g. we may see f(); but f is an
+          // import which does return a value, and we use that elsewhere. finalizeCalls
+          // fixes that up. what we do here is wherever a value is used, we set the right
+          // value, which is enough to ensure that the wasm ast is valid for such uses.
+          // this is important as we run the optimizer on functions before we get
+          // to finalizeCalls (which we can only do once we've read all the functions,
+          // and we optimize in parallel starting earlier).
           Ref parent = astStackHelper.getParent();
-          WasmType type = !!parent ? detectWasmType(parent, &asmData) : none;
-          noteImportedFunctionCall(ast, type, ret->cast<CallImport>());
+          import->type = !!parent ? detectWasmType(parent, &asmData) : none;
```

Comment:
My first thought: should this be wrapped with `if (import->type == none)` ?
Reading more closely, `import` is the call site, so this doesn't overwrite anything else.
Would it make sense to rename `import` to `callImport` to make it clearer?